iothub/service/tests/DigitalTwinsClientTests.cs

```diff
@@ -0,0 +1,468 @@
+
+            using var mockHttpResponse = new HttpResponseMessage
+            {
+                StatusCode = HttpStatusCode.NotFound,
+                Content = HttpMessageHelper.SerializePayload(responseMessage),
+            };
+
+            var mockHttpClient = new Mock<HttpClient>();
+            mockHttpClient
+                .Setup(restOp => restOp.SendAsync(It.IsAny<HttpRequestMessage>(), It.IsAny<CancellationToken>()))
+                .ReturnsAsync(mockHttpResponse);
+
+            var digitalTwinsClient = new DigitalTwinsClient(
+                HostName,
+                mockCredentialProvider.Object,
+                mockHttpClient.Object,
+                mockHttpRequestFactory,
+                s_retryHandler);
+
+            // act
+            Func<Task> act = async () => await digitalTwinsClient.GetAsync<string>(digitalTwinId);
+
+            // assert
+            await act.Should().ThrowAsync<IotHubServiceException>();
+        }
+
+        [TestMethod]
+        public async Task DigitalTwinsClient_UpdateAsync()
+        {
+            // arrange
+            string digitalTwinId = "foo";
+            var digitalTwin = new BasicDigitalTwin
+            {
+                Id = digitalTwinId,
+            };
+            var contents = new Dictionary<string, object>
+            {
+                { "temperature", 8 }
+            };
+            var jsonPatch = new JsonPatchDocument();
+            jsonPatch.AppendAdd("bar", contents);
+
+            var mockCredentialProvider = new Mock<IotHubConnectionProperties>();
+            mockCredentialProvider
+                .Setup(getCredential => getCredential.GetAuthorizationHeader())
+                .Returns(s_validMockAuthenticationHeaderValue);
+            var mockHttpRequestFactory = new HttpRequestMessageFactory(s_httpUri, "");
+
+            using var mockHttpResponse = new HttpResponseMessage
+            {
+                StatusCode = HttpStatusCode.Accepted,
+                Content = HttpMessageHelper.SerializePayload(digitalTwin),
+            };
+            mockHttpResponse.Headers.Add("ETag", "\"AAAAAAAAAAE=\"");
+            mockHttpResponse.Headers.Add("Location", "foo");
+            
+            var mockHttpClient = new Mock<HttpClient>();
+            mockHttpClient
+                .Setup(restOp => restOp.SendAsync(It.IsAny<HttpRequestMessage>(), It.IsAny<CancellationToken>()))
+                .ReturnsAsync(mockHttpResponse);
+
+            var digitalTwinsClient = new DigitalTwinsClient(
+                HostName,
+                mockCredentialProvider.Object,
+                mockHttpClient.Object,
+                mockHttpRequestFactory,
+                s_retryHandler);
+
+            // act
+            DigitalTwinUpdateResponse response = await digitalTwinsClient.UpdateAsync(digitalTwinId, jsonPatch.ToString());
+
+            // assert
+            response.Should().BeOfType<DigitalTwinUpdateResponse>();
```

Comment:
I don't know. Since we mock the response ourselves, wouldn't we have to manually serialize the new twin and we'd just be verifying a value we've manually set for the test?

Reply:
What I meant is we are already assigning the response to DigitalTwinUpdateResponse in line 183, so this assertion will not be reached if it failes in line 183.

Reply:
Yeah, I'm not seeing anything interesting that this test is validating. It gets code coverage but what have we learned, except that nothing exploded and that the return type is what we specified (how could it not be and still compile?).

I'll help. Let's chat on Teams and talk about it.

Reply:
Yup. You're both right. I got way too bogged down in the test itself and debugging it that I lost track of the actual value that it provides. I'll try and figure out a way to make this more useful.